File: typescript/src/value_presence.ts
import type { CaseSet, Category } from "./case_set_schema.ts";

const VALUE_PRESENCE_PASS_CATEGORIES = new Set<Category>([
  "correct",
  "wrong-subject",
  "stale-value",
  "hypothetical",
  "wrong-basis",
]);

export function formatClaimValue(value: number): string {
  const text = String(value);
  const dot = text.indexOf(".");
  const integerPart = dot === -1 ? text : text.slice(0, dot);
  const grouped = integerPart.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  return dot === -1 ? grouped : `${grouped}${text.slice(dot)}`;
}

export function valuePresent(text: string, formattedValue: string): boolean {
  const escaped = formattedValue.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`(?<![\\d.,])${escaped}(?!\\d|,\\d|\\.\\d)`).test(text);
}

export interface ValuePresenceResult {
  present: boolean;
  expectedPresent: boolean;
}

export function checkValuePresence(caseSet: CaseSet): Map<string, ValuePresenceResult> {
  const results = new Map<string, ValuePresenceResult>();
  for (const card of caseSet.cards) {
    const evidence = card.evidence_spans.map((span) => span.quote).join("\n");
    const present = valuePresent(evidence, formatClaimValue(card.claim.value));
    results.set(card.id, {
      present,
      expectedPresent: VALUE_PRESENCE_PASS_CATEGORIES.has(card.category),
    });
  }
  return results;
}

File: python/faai/value_presence.py
"""Naive value-presence matching for claim cards.

Mirrors typescript/src/value_presence.ts. This is the check the checker
experiment (#2) exists to beat: it passes any card whose value literally
appears in the evidence, including wrong-subject, stale, hypothetical and
wrong-basis cards. The case set is built so exactly those categories pass.
"""

from __future__ import annotations

import re
from typing import Any

VALUE_PRESENCE_PASS_CATEGORIES = frozenset(
    {"correct", "wrong-subject", "stale-value", "hypothetical", "wrong-basis"}
)


def format_claim_value(value: int | float) -> str:
    """Format a claim value the way the artifacts write figures: comma-grouped, minimal decimals."""
    if isinstance(value, float) and value.is_integer():
        value = int(value)
    return f"{value:,}"


def value_present(text: str, formatted_value: str) -> bool:
    """Naive value-presence: does the formatted figure appear, not inside a larger number?"""
    pattern = rf"(?<![\d.,]){re.escape(formatted_value)}(?!\d|,\d|\.\d)"
    return re.search(pattern, text) is not None


def check_value_presence(case_set: dict[str, Any]) -> dict[str, dict[str, bool]]:
    """Report naive value presence per card alongside the property's naive expectation."""
    results: dict[str, dict[str, bool]] = {}
    for card in case_set["cards"]:
        evidence = "\n".join(span["quote"] for span in card["evidence_spans"])
        present = value_present(evidence, format_claim_value(card["claim"]["value"]))
        results[card["id"]] = {
            "present": present,
            "expectedPresent": card["category"] in VALUE_PRESENCE_PASS_CATEGORIES,
        }
    return results
